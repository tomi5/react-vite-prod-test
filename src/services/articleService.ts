import axios from 'axios';
import { Article } from '../types/article';

const API_URL = 'https://jsonplaceholder.typicode.com';
const ITEMS_PER_PAGE = 10;

export const articleService = {
  getArticles: async (page: number): Promise<Article[]> => {
    const { data } = await axios.get<Article[]>(`${API_URL}/posts`, {
      params: {
        _page: page,
        _limit: ITEMS_PER_PAGE,
      },
    });
    return data;
  },

  getArticleById: async (id: number): Promise<Article | undefined> => {
    try {
      const { data } = await axios.get<Article>(`${API_URL}/posts/${id}`);
      return data;
    } catch (error) {
      return undefined;
    }
  },
};
