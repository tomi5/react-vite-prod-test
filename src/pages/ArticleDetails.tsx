import React, { useEffect } from 'react';
import { useParams, Link, useNavigate } from 'react-router-dom';
import { useQuery } from '@tanstack/react-query';
import { articleService } from '../services/articleService';

const ArticleDetails: React.FC = () => {
  const { id } = useParams();
  const navigate = useNavigate();

  const {
    data: article,
    isLoading,
    error,
  } = useQuery({
    queryKey: ['article', id],
    queryFn: () => articleService.getArticleById(Number(id)),
  });

  useEffect(() => {
    if (error) {
      navigate('/articles');
    }
  }, [error, navigate]);

  if (isLoading) {
    return <div className="text-center py-8">Ładowanie artykułu...</div>;
  }

  if (error || !article) {
    return null;
  }

  return (
    <div className="max-w-3xl mx-auto px-4">
      <Link to="/articles" className="text-blue-600 hover:text-blue-800 mb-6 inline-block">
        ← Wróć do listy artykułów
      </Link>
      <article className="bg-white rounded-lg shadow-md p-8">
        <h1 className="text-4xl font-bold mb-4">{article.title}</h1>
        <div className="prose max-w-none">
          {article.body.split('\n').map((paragraph, index) => (
            <p key={index} className="mb-4">
              {paragraph}
            </p>
          ))}
        </div>
      </article>
    </div>
  );
};

export default ArticleDetails;
