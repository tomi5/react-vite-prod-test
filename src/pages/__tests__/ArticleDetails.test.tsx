import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render } from '@testing-library/react';
import { screen, waitFor } from '@testing-library/dom';
import { MemoryRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import ArticleDetails from '../ArticleDetails';
import * as articleService from '../../services/articleService';
import '@testing-library/jest-dom';

vi.mock('../../services/articleService');
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual<typeof import('react-router-dom')>('react-router-dom');
  return {
    ...actual,
    useParams: () => ({
      id: '1',
    }),
    useNavigate: () => vi.fn(),
  };
});

describe('ArticleDetails', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  const mockArticle = {
    id: 1,
    title: 'Test Article',
    body: 'Test content',
    userId: 1,
    createdAt: '2024-03-20',
  };

  const renderWithProviders = (component: React.ReactElement) => {
    return render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>{component}</MemoryRouter>
      </QueryClientProvider>,
    );
  };

  beforeEach(() => {
    vi.clearAllMocks();
    queryClient.clear();
  });

  it('should display article details', async () => {
    vi.mocked(articleService.articleService.getArticleById).mockResolvedValue(mockArticle);

    const { container } = renderWithProviders(<ArticleDetails />);

    await waitFor(() => {
      expect(screen.getByText('Test Article')).toBeInTheDocument();
      expect(screen.getByText('Test content')).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it('should display error message for non-existent article', async () => {
    vi.mocked(articleService.articleService.getArticleById).mockResolvedValue(undefined);

    const { container } = renderWithProviders(<ArticleDetails />);

    await waitFor(() => {
      expect(screen.getByText('Ładowanie artykułu...')).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot while loading', () => {
    vi.mocked(articleService.articleService.getArticleById).mockImplementation(
      () => new Promise(() => {}), // never resolving promise to keep loading state
    );

    const { container } = renderWithProviders(<ArticleDetails />);

    expect(container).toMatchSnapshot();
  });
});
