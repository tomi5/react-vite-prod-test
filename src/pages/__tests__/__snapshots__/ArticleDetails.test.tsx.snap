// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticleDetails > should display article details 1`] = `
<div>
  <div
    class="max-w-3xl mx-auto px-4"
  >
    <a
      class="text-blue-600 hover:text-blue-800 mb-6 inline-block"
      href="/articles"
    >
      ← Powrót do artykułów
    </a>
    <article
      class="bg-white rounded-lg shadow-md p-8"
    >
      <h1
        class="text-4xl font-bold mb-4"
      >
        Test Article
      </h1>
      <div
        class="prose max-w-none"
      >
        <p
          class="mb-4"
        >
          Test content
        </p>
      </div>
    </article>
  </div>
</div>
`;

exports[`ArticleDetails > should display error message for non-existent article 1`] = `
<div>
  <div
    class="text-center py-8"
  >
    Ładowanie artykułu...
  </div>
</div>
`;

exports[`ArticleDetails > should match snapshot while loading 1`] = `
<div>
  <div
    class="text-center py-8"
  >
    Ładowanie artykułu...
  </div>
</div>
`;
